File: ts-core/src/net/link/frame.ts
import { Address } from "./address";
import { EnumSerdeable } from "@core/serde";

export enum Protocol {
    NoProtocol = 0,
    RoutingNeighbor = 1,
    RoutingReactive = 2,
    Rpc = 3,
    Observer = 4,
    Tunnel = 5,
}

export const PROTOCOL_LENGTH = 1;

export const uint8ToProtocol = (number: number): Protocol => {
    if (!(number in Protocol)) {
        throw new Error(`Unknown protocol number: ${number}`);
    }
    return number;
};

export const protocolToUint8 = (protocol: Protocol): number => {
    return protocol;
};

export const ProtocolSerdeable = new EnumSerdeable<Protocol>(Protocol);

export interface Frame {
    protocol: Protocol;
    remote: Address;
    payload: Uint8Array;
}

export const FRAME_MTU = 254;

File: py-core/py_core/net/link/frame.py
import asyncio
from dataclasses import dataclass
from enum import Enum
from typing import Final

from py_core.serde.enum import EnumSerde

from .address import Address


class ProtocolValue(Enum):
    NoProtocol = 0
    RoutingNeighbor = 1
    RoutingReactive = 2
    Rpc = 3
    Observer = 4
    Tunnel = 5


@dataclass(frozen=True)
class Protocol(EnumSerde[ProtocolValue]):
    @staticmethod
    def enum_type():
        return ProtocolValue


@dataclass
class LinkFrame:
    protocol: Protocol
    address: Address
    payload: bytes


@dataclass
class ReceivedLinkFrame(LinkFrame):
    media_disconnect_event: asyncio.Event


LINK_FRAME_MTU: Final[int] = 254


class PayloadTooLargeError(Exception):
    pass
